File: packages/sdk-python/examples/page_events.py
import asyncio
import json
import os

from pydantic import BaseModel

from stagehand import CDPSubscription, PageCDPEvent, Stagehand, browserbase

BROWSERBASE_API_KEY = os.environ.get("BROWSERBASE_API_KEY") or ""
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY") or ""
if not BROWSERBASE_API_KEY:
    raise RuntimeError("BROWSERBASE_API_KEY is required")
if not OPENAI_API_KEY:
    raise RuntimeError("OPENAI_API_KEY is required")


class PageInfo(BaseModel):
    heading: str
    description: str


async def main() -> None:
    browser = await browserbase.launch(
        api_key=BROWSERBASE_API_KEY,
        model="openai/gpt-5.4-mini",
        model_api_key=OPENAI_API_KEY,
    )
    try:
        stagehand = await Stagehand.create(browser=browser)
        subscription: CDPSubscription | None = None
        try:
            page = await stagehand.browser.context.active_page()
            if page is None:
                raise RuntimeError("Stagehand initialized without an active page")

            console_event = asyncio.Event()
            event_method: str | None = None

            def handle_console(event: PageCDPEvent) -> None:
                nonlocal event_method
                if event.params.root.get("type") == "log":
                    event_method = event.method
                    console_event.set()

            subscription = await page.on("console", handle_console)
            await page.goto("https://example.com")
            await page.evaluate('console.log("stagehand-page-on-example"); "emitted"')
            await asyncio.wait_for(console_event.wait(), timeout=10)

            result = await stagehand.extract(
                "Extract the page heading and description",
                PageInfo,
            )
            print(
                json.dumps(
                    {
                        "event_method": event_method,
                        "extracted": result.data.model_dump(mode="json"),
                    },
                    indent=2,
                )
            )
        finally:
            if subscription is not None:
                await subscription.unsubscribe()
            await stagehand.close()
    finally:
        await browser.close()


if __name__ == "__main__":
    asyncio.run(main())

File: packages/sdk-ruby/examples/page_events.rb
# frozen_string_literal: true

# Ruby port of packages/sdk-python/examples/page_events.py: subscribe to page
# console events via page.on, trigger one with evaluate, then AI-extract the
# page content.
#
#   ruby examples/page_events.rb                # local Chrome (needs OPENAI_API_KEY)
#   ruby examples/page_events.rb --browserbase  # Browserbase (Model Gateway)

require "json"

require_relative "../lib/stagehand"

PAGE_INFO = {
  "type" => "object",
  "properties" => {
    "heading" => { "type" => "string" },
    "description" => { "type" => "string" },
  },
  "required" => %w[heading description],
  "additionalProperties" => false,
}.freeze

openai_api_key = ENV.fetch("OPENAI_API_KEY", "")

browser =
  if ARGV.include?("--browserbase")
    puts "Creating a Browserbase session..."
    Stagehand::Browserbase.launch(api_key: ENV.fetch("BROWSERBASE_API_KEY"))
  else
    abort "Local runs need OPENAI_API_KEY (the Browserbase Model Gateway is session-only)." if openai_api_key.empty?
    puts "Launching local Chrome..."
    Stagehand::LocalBrowser.launch(headless: ENV["HEADED"].nil?)
  end

begin
  create_options = { browser: browser, log_level: ENV.fetch("STAGEHAND_LOG_LEVEL", "warn") }
  unless openai_api_key.empty?
    create_options[:model] = "openai/gpt-5.4-mini"
    create_options[:model_api_key] = openai_api_key
  end
  stagehand = Stagehand.create(**create_options)
  begin
    page = stagehand.browser.context.active_page()
    raise "Stagehand initialized without an active page" if page.nil?

    console_events = Thread::Queue.new
    subscription = page.on("console") do |event|
      # Runs on an SDK inbound thread (deliveries may be concurrent): hand
      # the event over to the queue rather than doing work here.
      console_events << event if event.params.is_a?(Hash) && event.params["type"] == "log"
    end

    begin
      page.goto("https://example.com")
      page.evaluate('console.log("stagehand-page-on-example"); "emitted"')
      event = console_events.pop(timeout: 10)
      raise "console event did not arrive within 10s" if event.nil?

      result = stagehand.extract("Extract the page heading and description", schema: PAGE_INFO)
      puts JSON.pretty_generate({ "event_method" => event.method, "extracted" => result.data })
    ensure
      subscription.unsubscribe
    end
  ensure
    stagehand.close
  end
ensure
  browser.close
end
